File: AutomaticDrawing.Painter.MouseAnimation/Painter/MousePainter.vb
Imports System.Numerics
Imports AutomaticDrawing.Core
Imports AutomaticDrawing.Utilities.Keyboard
''' <summary>
''' 提供鼠标控制的绘图器
''' </summary>
Public Class MousePainter
    Implements IPainter
    Public Event UpdatePaint As IPainter.UpdatePaintEventHandler Implements IPainter.UpdatePaint
    ''' <summary>
    ''' 鼠标事件间隔
    ''' </summary>
    Public Property SleepTime As Integer
    ''' <summary>
    ''' 绘制偏移
    ''' </summary>
    Public Property Offset As Vector2

    ''' <summary>
    ''' 创建并初始化一个实例
    ''' </summary>
    Public Sub New(offset As Vector2, Optional sleepTime As Integer = 3)
        Me.SleepTime = sleepTime
        Me.Offset = offset
    End Sub

    Public Sub Start(lines As List(Of ILine)) Implements IPainter.Start
        Dim totalCount As Integer = lines.SelectMany(Function(e As ILine)
                                                         Return e.Vertices
                                                     End Function).Count
        Dim current As Integer = 0
        For Each SubLine In lines
            If CheckKey() = False Then
                Return
            End If
            VirtualKeyboard.MouseMove(SubLine.Vertices.First.X + Offset.X, SubLine.Vertices.First.Y + Offset.Y, SleepTime)
            VirtualKeyboard.MouseDownOrUp(True, SleepTime)
            For Each SubPoint In SubLine.Vertices
                VirtualKeyboard.MouseMove(SubPoint.X + Offset.X, SubPoint.Y + Offset.Y, SleepTime)
                current += 1
                RaiseEvent UpdatePaint(Me, New UpdatePaintEventArgs(SubPoint, current / totalCount))
            Next
            VirtualKeyboard.MouseDownOrUp(False, SleepTime)
        Next
    End Sub
    Public Sub Pause() Implements IPainter.Pause
        Throw New NotImplementedException()
    End Sub
    Public Sub [Stop]() Implements IPainter.Stop
        Throw New NotImplementedException()
    End Sub

    Private Function CheckKey() As Boolean
        Dim key As Char = ChrW(VirtualKeyboard.GetActiveLetterKey())
        If key = StaticSource.HotKey_Pause Then
            Debug.WriteLine("Pause")
            Dim tempKey As Char
            While Not tempKey = StaticSource.HotKey_Continue
                tempKey = ChrW(VirtualKeyboard.GetActiveLetterKey())
                If tempKey = StaticSource.HotKey_Stop Then
                    Return False
                End If
                System.Threading.Thread.Sleep(10)
            End While
            Debug.WriteLine("Continue")
        ElseIf key = StaticSource.HotKey_Stop Then
            Debug.WriteLine("Stop")
            Return False
        End If
        Return True
    End Function
End Class

File: AutoPaint.Painter.MouseAnimation/Painter/MousePainter.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Linq;
using System.Numerics;
using System.Diagnostics;
using AutoPaint.Core;
using AutoPaint.Utilities;

namespace AutoPaint.Painter
{
    public class MousePainter : IPainter
    {
        public event EventHandler<OnPaintingUpdatedEventArgs> OnPaintingUpdated;

        public int SleepTime { get; set; }
        public Vector2 Offset { get; set; }
        public bool IsCheckCtrlAlKey { get; set; } = true;

        public MousePainter(Vector2 offset, int sleepTime = 3)
        {
            this.SleepTime = sleepTime;
            this.Offset = offset;
        }

        public void Start(IPainting painting)
        {
            var lines = painting.Lines;
            int totalCount = lines.SelectMany(e => e.Vertices).Count();
            int current = 0;
            foreach (var line in lines)
            {
                if (CheckKey() == false)
                    return;
                VirtualKeyboard.MouseMove((int)(line.Vertices.First().X + Offset.X), (int)(line.Vertices.First().Y + Offset.Y), SleepTime);
                VirtualKeyboard.MouseDown(SleepTime);
                foreach (var SubPoint in line.Vertices)
                {
                    VirtualKeyboard.MouseMove((int)(SubPoint.X + Offset.X), (int)(SubPoint.Y + Offset.Y), SleepTime);
                    current += 1;
                    OnPaintingUpdated?.Invoke(this, new OnPaintingUpdatedEventArgs(SubPoint, current / (float)totalCount));
                }
                VirtualKeyboard.MouseUp(SleepTime);
            }
        }
        public void Pause()
        {
            throw new NotImplementedException();
        }
        public void Resume()
        {
            throw new NotImplementedException();
        }
        public void Stop()
        {
            throw new NotImplementedException();
        }


        private bool CheckKey()
        {
            if ((!IsCheckCtrlAlKey) || (VirtualKeyboard.IsKeyDown(VirtualKeys.VK_CONTROL) && VirtualKeyboard.IsKeyDown(VirtualKeys.VK_MENU)))
            {
                char key = (char)VirtualKeyboard.GetActiveLetterKey();
                if (key == StaticResources.HotKey_Pause)
                {
                    Debug.WriteLine("Pause");
                    char? tempKey = null;
                    while (tempKey != StaticResources.HotKey_Continue)
                    {
                        tempKey = (char)VirtualKeyboard.GetActiveLetterKey();
                        if (tempKey == StaticResources.HotKey_Stop)
                            return false;
                        System.Threading.Thread.Sleep(10);
                    }
                    Debug.WriteLine("Continue");
                }
                else if (key == StaticResources.HotKey_Stop)
                {
                    Debug.WriteLine("Stop");
                    return false;
                }
            }
            return true;
        }
    }

}
